### ISSUE TITLE
v0.4.1 too: "AttributeError: module 'pynitrokey' has no attribute 'fido2'"

### ISSUE THREAD

**Issue by Alexus9** [Nitrokey/pynitrokey#47]
After installing the latest pynitrokey v0.4.1 for my Nitrokey Start, it cannot be called successfully:
```
$ nitropy
Traceback (most recent call last):
  File "/home/ge/.local/bin/nitropy", line 5, in <module>
    from pynitrokey.cli import nitropy
  File "/home/ge/.local/lib/python3.6/site-packages/pynitrokey/cli/__init__.py", line 15, in <module>
    import pynitrokey.fido2.operations
  File "/home/ge/.local/lib/python3.6/site-packages/pynitrokey/fido2/__init__.py", line 10, in <module>
    import pynitrokey.fido2.hmac_secret as hmac_secret
AttributeError: module 'pynitrokey' has no attribute 'fido2'
$
```
OS: openSUSE Leap 15.2
nitropy installed by
`pip3 install --user pynitrokey`
completed with message
`Successfully installed cbor-1.0.0 click-7.1.2 ecdsa-0.16.0 fido2-0.8.1 intelhex-2.2.1 pygments-2.7.1 pynitrokey-0.4.1 pyserial-3.4 pyusb-1.1.0`

What could be done? Thank you.

**Comment by daringer** [comment]
oookay, no chance to reproduce it until now, could you please check the following:
* is there a systemwide `pynitrokey` installed (output of `find /usr/lib/python3.8/site-packages/pynitrokey*` )
* please also paste me the output of your local install: `find ~/.local/lib/python3.8/site-packages/pynitrokey*`

my best guess is some old installation, partly related to #43.
another idea would be to open python directly and import the modules by hand, and check `sys.path` before:
```
$ python
Python 3.8.5 (default, Sep  5 2020, 10:50:12) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> list(sys.path)
['', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/home/daringer/.local/lib/python3.8/site-packages', '/usr/lib/python3.8/site-packages']
>>> import pynitrokey.fido2
>>> import pynitrokey.fido2 as f2
>>> f2.__file__
'/home/daringer/.local/lib/python3.8/site-packages/pynitrokey/fido2/__init__.py'
```

**Comment by Alexus9** [comment]
Thanks for the quick reply. Please find attached the output of the commands.
I tried also nitropy in a openSUSE Leap 15.2 live system, and I saw the same appearance.

[nitropy-check-20-10-05.txt](https://github.com/Nitrokey/pynitrokey/files/5325713/nitropy-check-20-10-05.txt)

**Comment by szszszsz** [comment]
@daringer Could this be connected in any way with older Python version (in this case 3.6)?

**Comment by daringer** [comment]
mmh, `.local` installation looks good, had a typo in my first line I asked you for, can you please paste `find /usr/lib64/python*/site-packages/pynitrokey*`, thanks.
Apart from this, I would not suspect a version issue between 3.6 and 3.8, but will try to reproduce it in a suse container, let's see...

**Comment by daringer** [comment]
looks fixed, with [75e567c](https://github.com/Nitrokey/pynitrokey/commit/75e567cc411e4fdec28994c78c6218bc84e8e382)

@Alexus9 could you confirm please, to install from git using `pip3` just do a: `pip3 install -U --user https://github.com/Nitrokey/pynitrokey/archive/master.zip` (or `uninstall` first, then leave out the `-U`) afterwards `nitropy` should work.

**Comment by szszszsz** [comment]
@daringer Interesting! Could you explain the difference in import calls?

**Comment by daringer** [comment]
apart from the obvious: using `from foo import bar` instead of `import foo.bar as bar`, honestly: not entirely ;)

* seen this before, fixed this before
* strongly suggesting it is more or less a circular-import issue
* behavior (and resolving) of these has been improved in recent python versions
* generally the structure of the current `pynk.fido2` is flawed, means e.g., `hmac_secret.py` needs `find` from `__init__.py` and `hmac_secret` is imported inside `__init__` for convenience (lazier imports) from other modules, by definition circular. Will be cleaned up towards 0.5 release

**Comment by Alexus9** [comment]
Yes, it works! Thank you very much!

```
$ nitropy version
*** Nitrokey tool for Nitrokey FIDO2 & Nitrokey Start
0.4.1
$
```

**Comment by daringer** [comment]
good to know, thx for the finding, 

hope you can survive with this some days...
will release a 0.4.2 the next days then a regular pip install should work for you then

**Comment by data-henrik** [comment]
FYI: I had trouble with the install from git. It gave an error during the actual install process. What worked was to edit the file  ~/.local/lib/python3.6/site-packages/pynitrokey/fido2/__init__.py and just replace the single line for the hmac_secret

**Comment by szszszsz** [comment]
@data-henrik Could you tell was it an update from the previous version, or a very first install of the pynitrokey package?

**Comment by data-henrik** [comment]
@szszszsz The first install failed which was the regular `pip3 install -U --user pynitrokey`. Then I found the info in this issue and tried to install it from git as well as uninstalling it first.

I had / have several FIDO modules installed before, e.g., Solo and Yubico.

**Comment by daringer** [comment]
> FYI: I had trouble with the install from git. It gave an error during the actual install process. What worked was to edit the file ~/.local/lib/python3.6/site-packages/pynitrokey/fido2/**init**.py and just replace the single line for the hmac_secret

Sorry, I do not fully understand, what exactly did not work, @data-henrik could you please give some more details, like:

> It gave an error during the actual install process.

Does this mean you encountered an error *during* the installation ? The error discussed in this issue is *during execution*. So, did you see the very same error as the *author* of this issue ?

> @szszszsz The first install failed which was the regular pip3 install -U --user pynitrokey.

Are you implying the same error here during installation, or during execution?

> What worked was to edit the file ~/.local/lib/python3.6/site-packages/pynitrokey/fido2/init.py and just replace the single line for the hmac_secret

So you reverted the commit: https://github.com/Nitrokey/pynitrokey/commit/75e567cc411e4fdec28994c78c6218bc84e8e382 from:
`from pynitrokey.fido2 import hmac_secret` to `import pynitrokey.fido2.hmac_secret as hmac_secret` and then it worked?

Finally, it is not entirely clear for me if it's working now for you (without any hand-modifications)?
Best-case you could also add your distribution + version 

Sorry for being so picky here, but without the details it's tough to write fixes ;)

**Comment by data-henrik** [comment]
Let me repeat :). First, I did not open a new defect because you already addressed it in a recent and yet to-be-released fix.

I ran into the same runtime issues (execution) as the originator of this issue. Thereafter, I tried to follow the advice provided here in the issue and tried to install from git:

```
$ pip3 install -U --user https://github.com/Nitrokey/pynitrokey/archive/master.zip
Collecting https://github.com/Nitrokey/pynitrokey/archive/master.zip
  Using cached https://github.com/Nitrokey/pynitrokey/archive/master.zip
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/usr/lib64/python3.6/tokenize.py", line 452, in open
        buffer = _builtin_open(filename, 'rb')
    FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-6ctl7l6e-build/setup.py'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-6ctl7l6e-build/
```
As can be seen, it resulted in an error. Thus, I resorted to applying the fix directly on my installed pynitrokey files:

`vi ~/.local/lib/python3.6/site-packages/pynitrokey/fido2/__init__.py`
Change:

`import pynitrokey.fido2.hmac_secret as hmac_secret`

 to

`from pynitrokey.fido2 import hmac_secret`

With that my installation works and does not give any runtime errors. I was able to update the firmware and run other fido2 commands.

Bottom line: My comment was added as FYI, no new issue needed, everything fine. ;-)

**Comment by daringer** [comment]
ahhh, ok much clearer now for me, especially the "bottom line" comment ;)
nevertheless it is weird that you see this error, looks like you cannot build the package according to [pep-517](https://www.python.org/dev/peps/pep-0517/) which might be due to an older `pip3` version, could you please re-try with an updated pip, because I would expect the pip install from git should be working without issues:
```
pip3 --version
python3 -m pip install -U pip
pip3 --version
pip3 install -U --user https://github.com/Nitrokey/pynitrokey/archive/master.zip
```
